sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mcps.c

```diff
@@ -412,7 +444,15 @@ void gnrc_lorawan_mcps_request(gnrc_lorawan_t *mac,
     gnrc_lorawan_build_uplink(mac, pkt, waiting_for_ack,
                               mcps_request->data.port);
 
+    if (mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_ADR_ANS) {
+        mac->mlme.pending_mlme_opts &= ~GNRC_LORAWAN_MLME_OPTS_LINK_ADR_ANS;
+        mac->mcps.adr_ack_cnt = 1;
```

Comment:
Okay :+1:

Reply:
`adr_ack_cnt ` has to be set to 1 after datarate has been configured by the NS ( `LinkADRAns` event), but the NS can send a  `LinkADRReq` even after the `LinkADRAns` event hence the value 1. I could reset to 0 but then code after ADR would need additional operations and would affect code readability.

Reply:
> adr_ack_cnt has to be set to 1 after DR has been configured on the node ( LinkADRAns event), but the NS can send a LinkADRReq even after the LinkADRAns event hence the value 1. I could reset to 0 but then code after ADR would need additional operations and would affect code readability.

From the standard:
```
If an end-device whose data rate is optimized by the network to usea data rate higher than its lowest available data rate,it periodically needs to validate that the network still receives the uplink  frames. Each  time  the  uplink  frame  counter  is  incremented  (for  each  new  uplink, repeated   transmissions do not increase   the   counter),the   device   increments an ADR_ACK_CNT counter.
```

The standards specifies that this value has to be incremented. Even if the NS can send a ADRReq immediately, this is send _after_ the transmission procedure. For the MAC layer it doesn't matter if this is sent in the same transmission procedure or later. That's why there's a PHY layer to abstract this (`gnrc_lorawan.c`, `gnrc_lorawan_region.c`).

> I could reset to 0 but then code after ADR would need additional operations and would affect code readability.

I personally don't understand why following the standard would affect code readability. As discussed offline 2 weeks ago, if you set this to one you would need the following operations:
- Set it to 1 there
- Set an increment to this value
- Reset it to zero (we don't want to "lie" to the upper layer not resetting this value).

Following the standard only requires 2 operations:
- Reset (initialization, when receiving a downlink)
- Increment (when sending a frame).

Did I miss something?

Reply:
> Set it to 1 there
> Set an increment to this value
>  Reset it to zero (we don't want to "lie" to the upper layer not resetting this value).

As stated this only happens while serving `LinkADRAns`. There are cases when NS responds with ` LinkADRReq` more than once, in that case instead of resetting it in MLME and incrementing in MCPS, the value was just set to 1.

I was of the impression that `ADRACKReq` is an MCPS event but it so happens now, after offline discussion, its better suited for MLME. I shall update the code accordingly.

Reply:
this hasn't been addressed

Reply:
Done.